Integration guide fix (#125)
spacemeshos/docs: docs/start/integration.md

:::info

> We are currently working on the new version of the Spacemesh API. It is currently available as `v2alpha` but we strongly recommend doing any new integrations with it. The old version `v1` will be deprecated afterwards. The API description is available [here](https://github.com/spacemeshos/api/tree/master/spacemesh/v2alpha1).

Reviewer: If we're "strongly recommending" that people integrate with the new API, it's time to remove "alpha" from the name. This is confusing.

Author: Agree. What should I change here?

Reviewer 2: We'd like to keep `alpha` to iterate quickly and then move to `beta` with 1.7.

Reviewer: I disagree, but this isn't the place to debate it

## Node Management​

### Running a Private Node

To start running a private node, first ensure that your computer meets (or better yet, exceeds) the minimum [system requirements](./smeshing/requirements.md). Then, you can follow the "running a local Full Node" section of the:

- [Basic Smeshing](./smeshing/smeshing_basic/smeshing.md/#running-a-local-full-node) guide to easily set up your node using Smapp (GUI), or the
- [Advanced Smeshing](./smeshing/smeshing_adv/setup.md/#running-a-local-full-node) guide to set up your node using the `go-spacemesh` (CLI).

> If you only need to run anNode for integration purposes, you can ignore the instructions for smeshing.

Reviewer: ```suggestion
If you only need to run a node for integration purposes, you can ignore the instructions for smeshing.
```

Now, there is only one canonical block per layer and all transactions in a layer have a strict ordering. Therefore, we use the terms "mesh" and "chain" interchangeably. The "mesh" still refers to the many proposals produced each layer, but these are ephemeral (read on) and are usually not relevant from the application layer perspective.

Besides, let's face it, "Spacemesh" sounds much cooler than "Spacechain". 😀

### Two Types of Data

The Spacemesh network consists of both **ephemeral** and **evergreen** data. Ephemeral data is gossiped and may be stored for some time, but is eventually pruned as it is not required to sync from genesis, reconstruct history, or verify the current state. Only archival nodes store ephemeral data forever. Block proposals (stored temporarily) and Hare messages (never stored) are ephemeral.

By contrast, evergreen data is kept forever. It is essential for reconstructing history and for verifying the current state of the mesh. Therefore, all nodes store it forever. Blocks and transactions are evergreen.

### Global State

> The global state is the current state of every Spacemesh account (i.e., its balance, code, counter value, and storage). The entire global state as of a current block or layer can be summarized in a single global state hash value (which two nodes can compare to see if they agree on the canonical state at any given point in time). Global state is never gossiped or shared among nodes (except as part of fast sync, which we have not built yet). Tge global state is calculated based on a node's view of history (i.e., every canonical block and transaction since genesis). Global state is therefore _implicit_ rather than _explicit._

Reviewer: ```suggestion
The global state is the current state of every Spacemesh account (i.e., its balance, code, counter value, and storage). The entire global state as of a current block or layer can be summarized in a single global state hash value (which two nodes can compare to see if they agree on the canonical state at any given point in time). Global state is never gossiped or shared among nodes (except as part of fast sync, which we have not built yet). The global state is calculated based on a node's view of history (i.e., every canonical block and transaction since genesis). Global state is therefore _implicit_ rather than _explicit._
```

### Time

> Time in Spacemesh is broken down into epochs which are two weeks long on the mainnet. An epoch contain layers that are five minutes long, and one epoch is exactly 4032 layers. Epochs and layers are both zero-indexed. The first two epochs (i.e., epochs 0 and 1, consisting of layers [0-8063]) were an initial genesis bootstrapping period that contained no blocks or transactions. Epochs and layers are uniquely identified by their index, and there can never be more than one with the same index. Reason being, it does not make sense to talk about two "different" layers with the same index, just as it does not make sense to talk about two "different" days with the same date.

Reviewer 3: ```suggestion
Time in Spacemesh is broken down into epochs two weeks long on the mainnet. An epoch contains layers that are five minutes long, and one epoch is exactly 4032 layers. Epochs and layers are both zero-indexed. The first two epochs (i.e., epochs 0 and 1, consisting of layers [0-8063]) were an initial genesis bootstrapping period that contained no blocks or transactions. Epochs and layers are uniquely identified by their index, and there can never be more than one with the same index. It does not make sense to talk about two "different" layers with the same index, just as it does not make sense to talk about two "different" days with the same date.
```

### Reorgs and Canonicity

Explained in our section on [blocks](./../learn/blocks.md). See [finality](#finality), below, for more in-depth information.

## Data Model

The Spacemesh data model is also different from the data model of other blockchains you may be familiar with.

### Spacemesh Accounts

Read our explainer [here](./../learn/accounts.md). Also, see [Accounts](#accounts), below, for more information.

### Epochs and Layers

As explained above, each epoch and each layer is a singleton. For example, there is (and will be) only ever one epoch 4 and one layer 9237. One epoch consists of (contains) 4032 layers.

### Blocks and Transactions

Read the explainer [here](./../learn/blocks.md#blocks-and-transactions).

### The Mesh

The "mesh" (a.k.a. "chain") consists of the canonical set of explicit, evergreen data (i.e., blocks and the transactions they contain) of all history back to genesis. The mesh typically only grows in append-only fashion, but in rare circumstances it can shrink due to a reorg. That is to say, some number of formerly-canonical blocks from the end of the chain may be removed and replaced with other blocks in case of a reorg.

### Activations

> Smeshers establish their eligibility to smesh by generating and submitting a [Proof of Space-Time (PoST)](../learn/post.md) to the network. This proof is contained in objects called ATXs ("activation transactions", though these are not actually transactions). See the [Advanced Smeshing Guide](./smeshing/smeshing_adv/post_init.md/#initialization) for more information.

Reviewer: ```suggestion
Smeshers establish their eligibility to smesh by generating and submitting a [Proof of Space-Time (PoST)](../learn/post.md) to the network. This proof is contained in objects called ATXs ("activation transactions", though these are not actually transactions, also known as "activations"). See the [Advanced Smeshing Guide](./smeshing/smeshing_adv/post_init.md/#initialization) for more information.
```

Author: Change accepted with additional changes.

Each ATX is linked to one smesher, is identified by that smesher's `smesherID` (a 32-byte public key), and targets one epoch. If a smesher generates multiple ATXs targeting the same epoch, that smesher becomes guilty of [equivocation](#malfeasance) and its identity will be disqualified.

ATXs have an associated [weight](../learn/atx.md/#weight) which depends on the contents of the smesher's PoST.

### Proposals and Rewards

Smeshers who successfully establish eligibility to smesh in a given epoch will be allotted one or more **eligibility slots** (which are layers). Every smesher gets at least one such slot per epoch. If a smesher is online and in sync when their slot arrives, they will generate and submit a **[block proposal]** to the network. The proposal contains a subset of the transactions present in the smesher's view of the **mempool** (the place where unconfirmed transactions live before being selected to be included in block proposals).

> These transactions are chosen in such a way that the union of all proposals from all smeshers is enough to efficiently fill a **block** (a set of confirmed transactions) with unique, valid transactions. In practice, until transaction throughput increases and blocks are regularly full, proposals will include the *entire* contents of a smesher's view of the mempool. The Hare consensus mechanism subsequently allows the network to agree on the canonical set of block proposals for a given layer, which then get deterministically assembled into a block for the layer.

Author: > 

Hmm, I think the original is still better. I want to bold "block" to signify it as an important concept.

> These transactions are chosen in such a way that the union of all proposals from all smeshers is enough to efficiently fill a **block** (a set of confirmed transactions) with unique, valid transactions. In practice, until transaction throughput increases and blocks are regularly full, proposals will include the *entire* contents of a smesher's view of the mempool. The Hare consensus mechanism subsequently allows the network to agree on the canonical set of block proposals for a given layer, which then get deterministically assembled into a block for the layer.

Reviewer 3: Sure but my change was mostly to put efficiently after "to fill a block" and not before the verb since efficiently is an adverb and should come after the verb.

> These transactions are chosen in such a way that the union of all proposals from all smeshers is enough to efficiently fill a **block** (a set of confirmed transactions) with unique, valid transactions. In practice, until transaction throughput increases and blocks are regularly full, proposals will include the *entire* contents of a smesher's view of the mempool. The Hare consensus mechanism subsequently allows the network to agree on the canonical set of block proposals for a given layer, which then get deterministically assembled into a block for the layer.

Author: What do you think about this:

"These transactions are chosen in such a way that the union of all proposals from all smeshers is enough to fill a **block** (a set of confirmed transactions) efficiently with unique, valid transactions."

> These transactions are chosen in such a way that the union of all proposals from all smeshers is enough to efficiently fill a **block** (a set of confirmed transactions) with unique, valid transactions. In practice, until transaction throughput increases and blocks are regularly full, proposals will include the *entire* contents of a smesher's view of the mempool. The Hare consensus mechanism subsequently allows the network to agree on the canonical set of block proposals for a given layer, which then get deterministically assembled into a block for the layer.

Reviewer 3: Yes looks good

:::info

We are currently working on the new version of the Spacemesh API. It is currently available as `v2alpha` but we strongly recommend doing any new integrations with it. The old version `v1` will be deprecated afterwards. The API description is available [here](https://github.com/spacemeshos/api/tree/master/spacemesh/v2alpha1).
There is also public instance exposing the JSON-RPC api at `mainnet-api.spacemesh.network`.

:::

This guide explains how to integrate Spacemesh into existing infrastructure, including block explorers, exchanges, and other types of applications. Note that libraries are still being worked on, including an [SDK](https://github.com/spacemeshos/spacemesh-sdk) in multiple programming languages like Go, Rust, and JavaScript. This is to make it easier for developers to integrate Spacemesh into their applications and perform actions such as talking to a Spacemesh node, generating addresses, and creating transactions. Until these libraries become available, this guide will provide you with all the information you need to write such code yourself and build applications on Spacemesh.

## Node Management​

> You can choose to either run a **full node** or a **smeshing node**. The former is a pre-requisite for the latter, and the difference between the two node types is that the full node only stores the blockchain state and executes the transactions in valid blocks, while the smeshing node performs smeshing as well.

Reviewer 2: I think this "full node" introduction here feels weird.

Author: Removed the introduction, reverted to "node" and "private node".

## Node Management​

You can choose to either run a **full node** or a **smeshing node**. The former is a pre-requisite for the latter, and the difference between the two node types is that the full node only stores the blockchain state and executes the transactions in valid blocks, while the smeshing node performs smeshing as well.

> ### Running a Private Node

Reviewer 2: That was actually meant to say Private here.

Author: Reverted to Private Node.

- [Basic Smeshing](./smeshing/smeshing_basic/smeshing.md/#running-a-local-full-node) guide to easily set up your node using Smapp (GUI), or the
- [Advanced Smeshing](./smeshing/smeshing_adv/setup.md/#running-a-local-full-node) guide to set up your full node using the `go-spacemesh` (CLI).

If you only need to run a Full Node for integration purposes, you can ignore the instructions for smeshing.

We encourage you to run multiple, redundant nodes so that you have a fallback in case one node has issues or falls out of sync. We also encourage you to take regular snapshots of the data directory of your nodes (ideally, after closing the node) so that you can restore your node more quickly in case of a problem, rather than needing to sync from scratch, which is time-consuming.

### Connecting to a Public Node

> Currently, we do not operate public nodes. However, we do plan to do so in the future. This depends upon improving the existing node API. You may follow [this issue](https://github.com/spacemeshos/pm/issues/250) for updates. For now, if you plan to build an application on Spacemesh, you will need to run your own nodes.

Reviewer 2: I think we can repeat here the v2alpha API info.

Author: Which info exactly? This one?

"We're currently working on the new version of the API. It's currently available as v2alpha but we strongly recommend doing any new integrations with it. The old version v1 will be deprecated afterwards. The API description is available [here](https://github.com/spacemeshos/api/tree/master/spacemesh/v2alpha1). There is also public instance exposing the JSON-RPC api at mainnet-api.spacemesh.network"

Reviewer 2: Hmm, some kind of like that I think.

Author: Added it. But I am not quite sure how v2alpha API info here is relevant for this section (Connecting to a Public Node).

#### Multiple Blocks

Unless security assumptions are violated, there will be zero or one block per layer. Examples of security assumption violations are a large number (> 1/3) of smesher weight colluding and equivocating, or a long-lived network partition (longer than an epoch) followed by a rejoin.

According to the Spacemesh protocol, only one block per layer may be considered canonical and only the transactions in that block may be included in the mesh. For more information on how this works in the protocol, see [explorer-backend#92](https://github.com/spacemeshos/explorer-backend/issues/92#issuecomment-1841293498).

Unfortunately, block validity is not currently exposed in the API (we're [working on it](https://github.com/spacemeshos/api/issues/287)). However, it can be manually read from the node database.

### Fetching Account Balance

> To get the balance of an account, use [`GlobalStateService.Account`](https://pkg.go.dev/github.com/spacemeshos/go-spacemesh@v1.2.1/api/grpcserver#GlobalStateService.Account). Balance is denominated in a **smidge**, where 1 SMH = 1 billion smidge (1e9 smidge). `stateCurrent` shows the current account balance while `stateProjected` includes transactions that have been mined into blocks but not yet applied to state (because the layer has not yet been finalized).

Reviewer: ```suggestion
To get the balance of an account, use [`GlobalStateService.Account`](https://pkg.go.dev/github.com/spacemeshos/go-spacemesh@v1.2.1/api/grpcserver#GlobalStateService.Account). Balance is denominated in **smidge**, where 1 SMH = 1 billion smidge (1e9 smidge). `stateCurrent` shows the current account balance while `stateProjected` includes transactions that have been mined into blocks but not yet applied to state (because the layer has not yet been finalized).
```

### Types

As mentioned above, Spacemesh implements native account abstraction: all accounts are represented by precompiled smart contracts known as "templates." There are no EOAs (i.e., accounts controlled by key pairs).

Spacemesh currently supports two types of transactions corresponding to the following two templates:

- Single-sig wallet: emulates an EOA and supports two methods, `Spawn` and `Spend`.
- Multisig wallet: same as Single-sig wallet but implements `m-of-n` key verification. Supports the same two methods, `Spawn` and `Spend`.

Once a transaction has been decoded as above, its type may be parsed by looking at the `template` and `method` fields. Here is a legend for mainnet transactions:

- SingleSig spawn: template `sm1qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqg56ypy7`, method `0` (or unset)
- SingleSig spend: template `sm1qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqg56ypy7`, method `16`
- MultiSig spawn: template `sm1qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqsl0g40s`, method `0` (or unset)
- MultiSig spend: template `sm1qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqsl0g40s`, method `16`

> Spend transactions have a destination (recipient) account and an amount. These are encoded inside the transaction's raw payload. This data is not currently exposed via the API but we're [working on it](https://github.com/spacemeshos/api/issues/288). For now, in order to fully parse a transaction, you will have to manually write code such as this [VM code](https://github.com/spacemeshos/go-spacemesh/blob/3d8eedfd44c6f9f8987e604b768596f6001fe703/genvm/templates/wallet/handler.go#L57-L72) that handles transaction parsing.

Reviewer: "data" is plural not singular :) https://www.britannica.com/dictionary/eb/qa/Is-Data-Singular-or-Plural-

Author: Thanks for the reference, it was definitely informative!

Hmm, it appears to be both as per the link, and I had changed it to its noncount form. But I will accept the change if you feel the plural form better suits the context.

Reviewer: you can use either, I'm being pedantic.

Note: there is one additional template: a Vault wallet, this is a special type of multisig wallet that is used to access the funds locked in the genesis vaults. However, vaulted funds do not begin to vest until one year post-genesis (i.e., in July 2024). Vault transactions should not appear before that time and will be invalid if they do so

### Lifecycle

The process of creating, broadcasting, and tracking a transaction looks something like the following:

> 1. Calculate the recipient account address, following the instructions above.

Reviewer: Please don't renumber lists. It's a feature of markdown that it performs the numbering for you automatically, which makes it much easier to update the list in the future.

Author: Did not know that! Reverted back to `1.`.
